drivers/network/dd/b57xx/ndis.c

```diff
@@ -0,0 +1,574 @@
+    NDIS_STATUS Status = NDIS_STATUS_PENDING;
+    PSCATTER_GATHER_LIST SGList = NDIS_PER_PACKET_INFO_FROM_PACKET(Packet,
+                                                                   ScatterGatherListPacketInfo);
+    LONG BusyDescriptors = Adapter->SendProducer[0].ProducerIndex -
+                           Adapter->SendProducer[0].ConsumerIndex;
+    
+    ASSERT(SGList != NULL);
+    ASSERT(SGList->NumberOfElements >= 1);
+    
+    NdisDprAcquireSpinLock(&Adapter->SendLock);
+    
+    if (BusyDescriptors < 0)
+    {
+        BusyDescriptors += Adapter->SendProducer[0].Count;
+    }
+    
+    if (Adapter->SendProducer[0].RingFull == TRUE ||
+        BusyDescriptors + SGList->NumberOfElements > Adapter->SendProducer[0].Count)
+    {
+        /* Force NDIS to resubmit the send */
+        Status = NDIS_STATUS_RESOURCES;
+        goto Exit;
+    }
+    
+    for (UINT i = 0; i < SGList->NumberOfElements; i++)
+    {
+        ASSERT(SGList->Elements[i].Length <= Adapter->MaxFrameSize);
+        
+        Adapter->SendProducer[0].pPacketList[Adapter->SendProducer[0].ProducerIndex] = Packet;
+
+        NICTransmitPacket(Adapter,
+                          SGList->Elements[i].Address,
+                          SGList->Elements[i].Length,
+                          i == SGList->NumberOfElements - 1 ? B57XX_SBD_PACKET_END : 0);
+    }
+    
+    if (Adapter->SendProducer[0].ProducerIndex == Adapter->SendProducer[0].ConsumerIndex)
+    {
+        NDIS_MinDbgPrint("Send ring is full\n");
+        Adapter->SendProducer[0].RingFull = TRUE;
+    }
+
+Exit:
+    NdisDprReleaseSpinLock(&Adapter->SendLock);
+    
+    return Status;
+}
+
+NTSTATUS
+NTAPI
+DriverEntry(_In_ PDRIVER_OBJECT DriverObject,
+            _In_ PUNICODE_STRING RegistryPath)
+{
+    NDIS_HANDLE WrapperHandle;
+    NDIS_STATUS Status;
+    NDIS_MINIPORT_CHARACTERISTICS Characteristics = {0};
+
+    Characteristics.MajorNdisVersion = NDIS_MINIPORT_MAJOR_VERSION;
+    Characteristics.MinorNdisVersion = NDIS_MINIPORT_MINOR_VERSION;
+    Characteristics.CheckForHangHandler = NULL;
+    Characteristics.DisableInterruptHandler = MiniportDisableInterrupt;
+    Characteristics.EnableInterruptHandler = MiniportEnableInterrupt;
+    Characteristics.HaltHandler = MiniportHalt;
+    Characteristics.HandleInterruptHandler = MiniportHandleInterrupt;
+    Characteristics.InitializeHandler = MiniportInitialize;
+    Characteristics.ISRHandler = MiniportISR;
+    Characteristics.QueryInformationHandler = MiniportQueryInformation;
+    Characteristics.ReconfigureHandler = NULL;
+    Characteristics.ResetHandler = MiniportReset;
+    Characteristics.SendHandler = MiniportSend;
+    Characteristics.SetInformationHandler = MiniportSetInformation;
+    Characteristics.TransferDataHandler = NULL;
+    Characteristics.ReturnPacketHandler = NULL;
+    Characteristics.SendPacketsHandler = NULL;
+    Characteristics.AllocateCompleteHandler = NULL;
+
```

Comment:
```suggestion
     Characteristics.AdapterShutdownHandler = MiniportShutdown;
```
Every NDIS 5.1 driver should have `MiniportShutdown` callback.